### ISSUE TITLE
TimeDistributed doesn't handle mask-reductions

### ISSUE THREAD

**Issue by braingineer** [fchollet/keras#2347]
A minimum working example: https://gist.github.com/braingineer/799d1d88a2b71b54691a983a55c727bd

The output:

```
==== input gets embedded =====
For x=X_emb
     x.ndim=5
     outmask.ndim=5
     inmask.ndim=5
==== a reduction from LSTM ====
For x=X_reduction, 
     x.ndim=4
     outmask.ndim=5
     inmask.ndim=5
```

in a nut shell, if the layer that TimeDistributed wraps reduces the number of dimensions and  there is a mask being passed through, the mask it not updated.  Down the line, this creates some issues (for me, it was that my y_pred reached the softmax, had the right `_keras_shape` but actually had `ndim=5` because I was using the mask at intermediate layers to keep the computations straight)

I'm not sure what a good fix would be.  Currently, I'm planning on hacking it with 

``` python
def compute_mask(self, x, mask=None):
    if mask is None:
        return None
    input_shape = self.input_spec[0].shape
    output_shape = self.get_output_shape_for(input_shape)
    if len(input_shape) > len(output_shape): 
        mask = K.max(mask, axis=-1)
    return mask
```

This requires keeping the output mask of layers consistent: whatever the output shape of a layer is going to be, the mask should be that size.  I've just been adding a new axis.  For example,  the Embedding layer:

``` python
    def compute_mask(self, x, mask=None):
        if not self.mask_zero:
            return None
        else:
            return K.expand_dims(K.not_equal(x, 0), -1)
```

Though, I'm not sure if that's all the right way to go.  Masking can be confusing.

**Comment by fchollet** [comment]
The current behavior of the `TimeDistributed` layer is to pass the mask it receives, unchanged. But you raise a good point. One reasonable behavior to implement would be to run each temporal slice of the mask to the child layer, then concatenate the resulting masks (only when a non-None mask is received, of course). Would you like to prepare a PR?

**Comment by braingineer** [comment]
Sure.  It's deadline season so it might be a couple of days. Also, it'll be my first PR ever, so please let me know if there's any issues/inaccuracies/things that could be tightened up.  

By the way, I wrote wrapper-like class that distributes over the last 2 dimensions of a tensor of 4 dimensions or larger with the assumption that the provided layer will reduce the last 2 dimensions to 1 (such as an RNN that only returns the last output).   I'm not sure if this is something you would like contributed as well.  As of now, the layer is named Summarize.
